eras/alonzo/impl/testlib/Test/Cardano/Ledger/Alonzo/Imp/UtxowSpec.hs

```diff
@@ -0,0 +1,168 @@
+import Cardano.Ledger.BaseTypes
+import Cardano.Ledger.Core
+import Cardano.Ledger.Plutus
+import Cardano.Ledger.Shelley.LedgerState (esLStateL, lsUTxOStateL, nesEsL, utxosUtxoL)
+import Cardano.Ledger.Shelley.Rules (ShelleyUtxowPredFailure (..))
+import Cardano.Ledger.UTxO (UTxO (..))
+import qualified Data.Map as Map
+import Lens.Micro
+import qualified PlutusLedgerApi.Common as P
+import Test.Cardano.Ledger.Allegra.Imp.UtxowSpec (genInvalidMetadatum)
+import Test.Cardano.Ledger.Allegra.ImpTest (produceScript)
+import Test.Cardano.Ledger.Alonzo.ImpTest (AlonzoEraImp, fixupPPHash)
+import Test.Cardano.Ledger.Imp.Common
+import Test.Cardano.Ledger.Plutus.Examples (guessTheNumber3)
+import Test.Cardano.Ledger.Shelley.ImpTest
+
+spec ::
+  forall era.
+  ( AlonzoEraImp era
+  , TxAuxData era ~ AlonzoTxAuxData era
+  , InjectRuleFailure "LEDGER" ShelleyUtxowPredFailure era
+  , InjectRuleFailure "LEDGER" AlonzoUtxowPredFailure era
+  , InjectRuleFailure "LEDGER" AlonzoUtxosPredFailure era
+  , Arbitrary (Data era)
+  ) =>
+  SpecWith (ImpTestState era)
+spec = describe "UTXOW" $ do
+  it "MissingRedeemers" $ do
+    let lang = eraMaxLanguage @era
+    let scriptHash = withSLanguage lang (hashPlutusScript . guessTheNumber3)
+    txIn <- produceScript scriptHash
+    let txToSubmit = mkBasicTx mkBasicTxBody & bodyTxL . inputsTxBodyL .~ [txIn]
+    let fixup tx =
+          fixupResetAddrWits $ tx & witsTxL . rdmrsTxWitsL .~ Redeemers mempty
+    let missingRedeemer = mkSpendingPurpose $ AsItem txIn
+    withPostFixup fixup $
+      submitFailingTx
+        txToSubmit
+        [ injectFailure $
+            MissingRedeemers [(missingRedeemer, scriptHash)]
+        , injectFailure $
+            CollectErrors [NoRedeemer missingRedeemer]
+        ]
+
+  it "MissingRequiredDatums" $ do
+    let lang = eraMaxLanguage @era
+    let scriptHash = withSLanguage lang (hashPlutusScript . guessTheNumber3)
+    txIn <- produceScript scriptHash
+    let txToSubmit = mkBasicTx mkBasicTxBody & bodyTxL . inputsTxBodyL .~ [txIn]
+    let fixup tx =
+          fixupResetAddrWits $ tx & witsTxL . datsTxWitsL .~ mempty
+    let missingDatum = hashData @era (Data (P.I 3))
+    withPostFixup fixup $
+      submitFailingTx
+        txToSubmit
+        [ injectFailure $
+            MissingRequiredDatums [missingDatum] []
+        ]
+
+  it "NotAllowedSupplementalDatums" $ do
+    let lang = eraMaxLanguage @era
+    let scriptHash = withSLanguage lang (hashPlutusScript . guessTheNumber3)
+    txIn <- produceScript scriptHash
+    let extraDatumHash = hashData @era (Data (P.I 30))
+    let extraDatum = Data (P.I 30)
+    let tx =
+          mkBasicTx mkBasicTxBody
+            & bodyTxL . inputsTxBodyL .~ [txIn]
+            & witsTxL . datsTxWitsL .~ TxDats (Map.singleton extraDatumHash extraDatum)
+    submitFailingTx
+      tx
+      [ injectFailure $
+          NotAllowedSupplementalDatums [extraDatumHash] []
+      ]
+
+  it "PPViewHashesDontMatch" $ do
```

Comment:
Let's get rid of:
* https://github.com/IntersectMBO/cardano-ledger/blob/2c9ba10d085331fa99840a63b40b8e8b716fba5d/libs/cardano-ledger-test/src/Test/Cardano/Ledger/Examples/AlonzoInvalidTxUTXOW.hs#L185-L207
* https://github.com/IntersectMBO/cardano-ledger/blob/2c9ba10d085331fa99840a63b40b8e8b716fba5d/libs/cardano-ledger-test/src/Test/Cardano/Ledger/Examples/AlonzoInvalidTxUTXOW.hs#L448-L462